x-pack/plugins/security_solution/server/lib/telemetry/tasks/diagnostic.ts

```diff
@@ -35,29 +41,48 @@ export function createTelemetryDiagnosticsTaskConfig() {
           throw new Error('last execution timestamp is required');
         }
 
-        const response = await receiver.fetchDiagnosticAlerts(
-          taskExecutionPeriod.last,
-          taskExecutionPeriod.current
-        );
+        let alertCount = 0;
+        let fetchMore = true;
+        let searchAfterValue: SortResults | undefined;
+        let pitId = await receiver.openPointInTime(DEFAULT_DIAGNOSTIC_INDEX);
```

Comment:
Perhaps instead of exposing the `pit` logic outside the function, it would be better to create the pit internally and return it as an opaque value so that the caller doesn't need to worry about these implementation details, WDYT?

Another possibility would be to use generators, pseudocode:

```node
function* fetchDiagnosticAlertsBatch() {
  let pit = createPit();

  while (fetchData) {
    // search...
    yield ...;
  }
  closePit();
}

for (const item of fetchDiagnosticAlertsBatch()) {
  ...
}
```

Reply:
I will spend today thinking about this. When you say _better_, what do you mean?

Reply:
better from a design perspective, to not expose internal details to the function's user.  I mean, as an API user I want to get data, and not worry about details like creating and closing a PIT.  Instead of 

```node
   pitId = ...
  while (moreToFetch) {
   const { moreToFetch, newPitId, searchAfter, alerts } =
            await receiver.fetchDiagnosticAlertsBatch(
              taskExecutionPeriod.last,
              taskExecutionPeriod.current,
              pitId,
              searchAfterValue
            );
    ....
   sendEvents( .. )
  }
  close(pitId)
```
IMHO it simpler and clearer

```node
  for  (const event of fetchDiagnosticAlertsBatch(last, current) {
   send(event) 
 }
```

Reply:
Thanks! Updated in https://github.com/elastic/kibana/pull/172087/commits/765f1d1db065b69050bd63e666a57ddb195728c9